core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -288,13 +288,36 @@ class half_t {
     return *this;
   }
 
+  template <class T>
+  KOKKOS_FUNCTION void operator=(T rhs) volatile {
+    val = cast_to_half(rhs).val;
+  }
+
   // Compound operators
   KOKKOS_FUNCTION
   half_t& operator+=(half_t rhs) {
 #ifdef __CUDA_ARCH__
     val += rhs.val;
 #else
-    val     = __float2half(__half2float(val) + __half2float(rhs.val));
+    val = __float2half(__half2float(val) + __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  volatile half_t& operator+=(half_t rhs) volatile {
+#ifdef __CUDA_ARCH__
+    // Cuda 10 supports __half volatile stores but not volatile arithmetic
+    // operands. Cast away volatile-ness of val for arithmetic but not for store
+    // location.
+    val = const_cast<impl_type*>(&val)[0] + rhs.val;
```

Comment:
```suggestion
    val = const_cast<impl_type&>(val) + rhs.val;
```

Reply:
OK. Leaving as-in since both are functionally equivalent.

Reply:
I agree with damien here. Its a significant clarity improvement.